mempool/mempool.go

```diff
@@ -270,10 +304,19 @@ func (mem *Mempool) filterTxs(blockTxsMap map[string]struct{}) []types.Tx {
 	goodTxs := make([]types.Tx, 0, mem.txs.Len())
 	for e := mem.txs.Front(); e != nil; e = e.Next() {
 		memTx := e.Value.(*mempoolTx)
+		// Remove the tx if its alredy in a block.
 		if _, ok := blockTxsMap[string(memTx.tx)]; ok {
-			// Remove the tx since already in block.
+			// remove from clist
 			mem.txs.Remove(e)
 			e.DetachPrev()
+
+			// remove from mempool cache
+			// we only enforce "at-least once" semantics and
+			// leave it to the application to implement "only-once"
+			// via eg. sequence numbers, utxos, etc.
+			// NOTE: expects caller of filterTxs to hold the lock
+			// (so we can't use mem.removeTxFromCacheMap)
+			delete(mem.cacheMap, string(memTx.tx))
```

Comment:
I'd rather keep the delete and let the app handle the case where a peer broadcasts the tx late. Otherwise, the cache will always eventually fill up with transactions that have been committed!